proposals/extensions_v2.md

```diff
@@ -0,0 +1,401 @@
+
+## Summary
+[summary]: #summary
+
+Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.
+
+This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).
+
+Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with classic extension methods, and then potential enhancements on top of this kernel to make certain common designs more succinct and pleasant.  This does not preclude the feature shipping with any or all of these enhancements at launch.  It simply allows the design to be broken into separable concerns, with a clearer ordering of dependencies.
+
+A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:
+
+Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  For example, a form that is only good for properties, but not for other members, it likely not appropriate.
+
+```c#
+extension E
+{
+    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
+    public int M<X>(...) for SomeType<X> val { }
+
+    // Property form:
+    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
+    public int Count<X> for SomeType<X> val { get { ... } }
+
+    // Event form:
+    // Note: would have to be the add/remove form.
+    // field-backed events would not be possible as extensions do not have instance state.
+    public event Action E<X> for SomeType<X> val { add { } remove { } }
+
+    // Indexer form:
+    public int this<T>[int index] for SomeType<X> val { get { ... } }
+
+    // Operator form:
+    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
+    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
+    
+    // Conversion form:
+    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
+    public static implicit operator SomeTypeX<X>(int i) for SomeType<X> { ... }
+
+    // Constructor form:
+    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
+    public SomeType<X>() for SomeType<X> { }
+
+    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
+    // note: no `Assert val`, a static method is not passed an instance value.
+    public static bool Boo(string s1, string s2) for Assert { }
+
+    // Static extensions properties, indexers and events are all conceptually supportable.
```

Comment:
>indexers [](http://example.com/codeflow?start=37&length=8)

I expect we'd support static extension properties, but not static extension indexers, as static indexers are not allowed (yet)